plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DefaultServerEvaluationStrategy.java

```diff
@@ -0,0 +1,253 @@
+     *      internal container address obtained from {@link ContainerInfo} e.g. via
+     *      {@code ContainerInfo.getNetworkSettings().getIpAddress()}</li>
+     * <li> If neither of the above properties are set, the address of the Docker daemon is used, as obtained
+     *      from {@link ContainerInfo} via {@code ContainerInfo.getNetworkSettings().getGateway()</li>
+     */
+    private String internalAddress;
+
+    /**
+     * External address of the server. In order of precedence, it is set according to
+     * <li> If property {@code che.docker.ip.external} is not null, that value is used.</li>
+     * <li> If property {@code che.docker.ip} is not null, that value is used.</li>
+     * <li> If neither of the above are true, then the value of the Docker daemon is used, as obtained
+     *      from {@link ContainerInfo}</li>
+     */
+    private String externalAddress;
+
+    /**
+     * Switch that stores whether we are using an internal container address. It is required because
+     * in the case where we are using direct communication between containers (i.e. within the docker0 network),
+     * we need to communicated on the ports exposed by the container.
+     */
+    private boolean useEphemeralPorts;
+
+    @Inject
+    public DefaultServerEvaluationStrategy(@Assisted ContainerInfo containerInfo,
+                                           @Assisted Map<String, ServerConfImpl> serverConf,
+                                           @Assisted String internalHost,
+                                           @Nullable @Named("che.docker.ip") String internalAddress,
+                                           @Nullable @Named("che.docker.ip.external") String externalAddress,
+                                           @Nullable @Named("che.docker.ip.use_internal_address") boolean useInternal) {
+        this.serverConf       = serverConf;
+
+        if (containerInfo.getNetworkSettings() != null && containerInfo.getNetworkSettings().getPorts() != null) {
+            ports = containerInfo.getNetworkSettings().getPorts();
+        } else {
+            ports = Collections.emptyMap();
+        }
+
+        if (containerInfo.getConfig() != null && containerInfo.getConfig().getLabels() != null) {
+            labels = containerInfo.getConfig().getLabels();
+        } else {
+            labels = Collections.emptyMap();
+        }
+
+        // If containerInfo did not contain ports, we will not be able to return servers. Instead,
+        // return early to avoid having to constantly check for null.
+        if (ports.isEmpty()) {
+            return;
+        }
+
+        // Make sure we set internal/external address according to precedence (see doc)
+        String containerAddress = containerInfo.getNetworkSettings().getIpAddress();
+        String dockerAddress    = containerInfo.getNetworkSettings().getGateway();
+        this.useEphemeralPorts = true;
+
+        if (internalAddress != null) {
```

Comment:
If we refactor code in such a way that common code would go to some abstract class or something like that strategies would be super simple. Probably in that case default strategy can have logic of ephemeral/exposed port choice.